sklearn/gaussian_process/gpr.py

```diff
@@ -0,0 +1,148 @@
+"""Gaussian processes regression. """
+
+# Authors: Jan Hendrik Metzen <jhm@informatik.uni-bremen.de>
+#
+# License: BSD 3 clause
+
+import numpy as np
+from scipy.linalg import cholesky, cho_solve, solve
+from scipy.optimize import fmin_l_bfgs_b
+
+from sklearn.base import BaseEstimator
+
+
+class GaussianProcessRegression(BaseEstimator):
+    """ Gaussian process regression (GPR).
+
+    The implementation is based on Algorithm 2.1 of ``Gaussian Processes
+    for Machine Learning'' (GPML) by Rasmussen and Williams.
+
+    In addition to standard sklearn estimators, GaussianProcessRegression
+       * allows prediction without prior fitting (based on the GP prior)
+       * provides an additional method sample(X), which evaluates samples drawn
+         from the GPR (prior or posterior) at given inputs
+       * exposes a method log_marginal_likelihood(theta), which can be used
+         externally for other ways of selecting hyperparamters, e.g., via
+         Markov chain Monte Carlo.
+
+    Parameters
+    ----------
+    kernel : Kernel object
+        The kernel specifying the covariance function of the GP.
+
+    y_err : float, optional (default: 1e-10)
+        Value added to the diagonal of the kernel matrix during fitting.
+        Larger values correspond to increased noise level in the observations
+        and reduce potential numerical issue during fitting.
+
+    Attributes
+    ----------
+    X_fit_:
+
+    y_fit_:
+
+    theta_:
+
+    L_:
+
+    alpha_:
+    """
+
+    def __init__(self, kernel, y_err=1e-10):
+        self.kernel = kernel
+        self.y_err = y_err
+
+    def fit(self, X, y):
+        self.X_fit_ = np.asarray(X)
+        self.y_fit_ = np.asarray(y)
+
+        if self.kernel.has_bounds:
+            # Choose hyperparameters based on maximizing the log-marginal
+            # likelihood
+            def obj_func(theta):
+                lml, grad = self.log_marginal_likelihood(theta,
+                                                         eval_gradient=True)
+                return -lml, -grad
+            self.theta_, lml, _ = fmin_l_bfgs_b(obj_func, self.kernel.params,
+                                                bounds=self.kernel.bounds)
+            self.kernel.params = self.theta_
+        else:
+            self.theta_ = self.kernel.params
+
+        # Precompute quantities required for predictions which are independent
+        # of actual query points
+        K = self.kernel.auto(self.X_fit_)
+        K[np.diag_indices_from(K)] += self.y_err
+        self.L_ = cholesky(K, lower=True)  # Line 2
+        self.alpha_ = cho_solve((self.L_, True), self.y_fit_)  # Line 3
```

Comment:
I would prefer this to be called `self.dual_coef_` or `self.coef_` as is the case for many other estimators.